70_FLIPFluids/app_resources/gridSampling.hlsl

```diff
@@ -0,0 +1,108 @@
+#ifndef _FLIP_EXAMPLE_GRID_SAMPLING_HLSL
+#define _FLIP_EXAMPLE_GRID_SAMPLING_HLSL
+
+#include "gridUtils.hlsl"
+
+#ifdef __HLSL_VERSION
+
+// adapted from CUDA Cubic B-Spline Interpolation by Danny Ruitjers
+// https://www.dannyruijters.nl/cubicinterpolation/
+inline float4 cubic(float v)
+{
+    float4 n = float4(1.0f, 2.0f, 3.0f, 4.0f) - v;
+    float4 s = n * n * n;
+    float x = s.x;
+    float y = s.y - 4.0f * s.x;
+    float z = s.z - 4.0f * s.y + 6.0f * s.x;
+    float w = 6.0 - x - y - z;
+    return float4(x, y, z, w) * (1.0f / 6.0f);
+}
+
+inline float4 tricubicInterpolate(Texture3D<float4> grid, SamplerState _sampler, float3 coords, int4 gridSize)
+{
+    float4 texelSize = float4(1.0f / gridSize.xz, gridSize.xz);
+    coords = coords * gridSize.xyz - 0.5f;
+
+    float3 invCoords = frac(coords);
+    coords -= invCoords;
+
+    float4 xcubic = cubic(invCoords.x);
+    float4 ycubic = cubic(invCoords.y);
+    float4 zcubic = cubic(invCoords.z);
+
+    float2 cx = coords.xx + float2(-0.5f, 1.5f);
+    float2 cy = coords.yy + float2(-0.5f, 1.5f);
+    float2 cz = coords.zz + float2(-0.5f, 1.5f);
+
+    float2 sx = xcubic.xz + xcubic.yw;
+    float2 sy = ycubic.xz + ycubic.yw;
+    float2 sz = zcubic.xz + zcubic.yw;
+
+    float2 offsetx = cx + xcubic.yw / sx;
+    float2 offsety = cy + ycubic.yw / sy;
+    float2 offsetz = cz + zcubic.yw / sz;
+    offsetx /= gridSize.xx;
+    offsety /= gridSize.yy;
+    offsetz /= gridSize.zz;
+
+    float4 tex0 = grid.SampleLevel(_sampler, float3(offsetx.x, offsety.x, offsetz.x), 0);
+    float4 tex1 = grid.SampleLevel(_sampler, float3(offsetx.y, offsety.x, offsetz.x), 0);
+    float4 tex2 = grid.SampleLevel(_sampler, float3(offsetx.x, offsety.y, offsetz.x), 0);
+    float4 tex3 = grid.SampleLevel(_sampler, float3(offsetx.y, offsety.y, offsetz.x), 0);
+    float4 tex4 = grid.SampleLevel(_sampler, float3(offsetx.x, offsety.x, offsetz.y), 0);
+    float4 tex5 = grid.SampleLevel(_sampler, float3(offsetx.y, offsety.x, offsetz.y), 0);
+    float4 tex6 = grid.SampleLevel(_sampler, float3(offsetx.x, offsety.y, offsetz.y), 0);
+    float4 tex7 = grid.SampleLevel(_sampler, float3(offsetx.y, offsety.y, offsetz.y), 0);
+
+    float gx = sx.x / (sx.x + sx.y);
+    float gy = sy.x / (sy.x + sy.y);
+    float gz = sz.x / (sz.x + sz.y);
+
+    float4 x0 = lerp(tex1, tex0, gx);
+    float4 x1 = lerp(tex3, tex2, gx);
+    float4 x2 = lerp(tex5, tex4, gx);
+    float4 x3 = lerp(tex6, tex6, gx);
+
+    float4 y0 = lerp(x1, x0, gy);
+    float4 y1 = lerp(x3, x2, gy);
+    
+    return lerp(y1, y0, gz);
+}
```

Comment:
great Google-FU

Reply:
P.S. `inline` means nothing in HLSL, only place its useful is `C++/HLSL` mixed code

Reply:
btw because of https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/pull/145/files#r1796660152 this function should be "monochrome" i.e. work in `float32_t` instead of `float32_t4`